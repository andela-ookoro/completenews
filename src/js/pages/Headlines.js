import React from 'react';
import ReactDOM from 'react-dom';
//import Fs from 'fs';
//import Request from 'request';
//import Cheerio from 'cheerio';
// import HealineStore from '../store/HeadlineStore';


class Headlines extends React.Component {
  constructor() {
    super();
    this.state = { source: '', sources: [], articles: [], articleSource: '', sortBy: [] };
    this.fecthHealines = this.fecthHealines.bind(this);
    // this.scrape = this.scrape.bind(this);
  }
  // this method runs before the component render it content
  componentWillMount() {
    this.getSources();
    // HealineStore.on("change", this.getSources);
    // console.log("count" ,HealineStore.listenerCount("change"));
  }

  componentWillUnmount() {
    // HealineStore.removeListener("change", this.getSources);
  }

  getSources() {
    if (!localStorage.sources) {
      return $.getJSON('https://newsapi.org/v1/sources?language=en')
       .then((response) => {
         this.setState({ sources: response.sources });
         localStorage.sources = [];
         localStorage.sources = JSON.stringify(response.sources);
         //console.log(this.state.sources);
         //console.log(localStorage.sources);
        });
    } 
    return this.setState({ sources: JSON.parse(localStorage.sources) });
  }

  toTitleCase(str)  {
    return str.replace(/\w\S*/g, txt => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase());
  }

  // get available sort parameter
  fetchAvailableSort() {
    const source = ReactDOM.findDOMNode(this.refs.sources).value;
    // fix source name
    let sourceName = this.toTitleCase(source.toString().replace(/-/g, ' ')),
      sources = JSON.parse(localStorage.sources),
      sourceNode = sources.filter(obj => obj.id == source);
    const apiURl = `https://newsapi.org/v1/articles?source=${source
            }&apiKey=213327409d384371851777e7c7f78dfe`;
    $.getJSON(apiURl)
      .then((response) => {
        this.setState({ source, sortBy: (sourceNode[0].sortBysAvailable.length > 1) ? sourceNode[0].sortBysAvailable : [], articleSource: sourceName, articles: response.articles });
      });

    // this.setState({ source:source, sortBy: sourceNode[0].sortBysAvailable, articleSource:sourceName});
  }

  // fetct headlines
  fecthHealines(e) {
    const sort = e.target.value;
    console.log(sort);
    const source = ReactDOM.findDOMNode(this.refs.sources).value;
    if (source) {
      const apiURl = `https://newsapi.org/v1/articles?source=${source
            }&sortBy=${sort
            }&apiKey=213327409d384371851777e7c7f78dfe`;
      $.getJSON(apiURl)
      .then((response) => {
            this.setState({ articles: response.articles });
          });
    }
  }

  scrape() {
    const url = 'https://thenextweb.com/microsoft/2017/05/02/microsoft-bringing-mixed-reality-classroom-view-mixed-reality/';
    Request(url, (error, response, html) => {
    if (!error) {
      Fs.writeFile('output.html', html, (err) => {
              console.log('File successfully written! - Check your project directory for the output.json file');
      });
    }
  });
  }


  render() {
    const sourceStyle = {
          marginleft: '5px',
          float: 'left',
    };
    const articleStyle = {
          marginleft: '5px',
          float: 'right',
    };
    const clear = {
      clear: 'both',
    };

    return (
      <div class="row"> 
        <div class="col s2">
          Sources
          <select style={{'minHeight': '700px'}} class="browser-default" name="sources" size={"25"} ref="sources" onChange={this.fetchAvailableSort.bind(this)}>
                {this.state.sources.map((source, i) => <SourceOptions key={i} data={source} />)}
            </select>
        </div>
        <div class="col s6">
          <h6>
            {this.state.articleSource} 
            {this.state.sortBy.map((sortBY, i) => <SortBY key={i} data={sortBY} onClick={this.fecthHealines} />)}
              </h6>
              {this.state.articles.map((article, i) => <Article key={i} data={article} />)}
        </div>
    </div>
    );
  }
}

// class to display source
class SourceOptions extends React.Component {
  render() {
    return (
      <option value={this.props.data.id}>{this.props.data.name}</option>
    );
  }
}

// class to display sortBy
class SortBY extends React.Component {
  render() {
      // console.log(this.props.onClick);
    return (
      <button value={this.props.data} onClick={this.props.onClick}> {this.props.data}            </button>

    );
  }
}

// class to show article
class Article extends React.Component {
  render() {
      const imgStyle = {
          width: '80%',
          height: '40%',
   };
    const paragraphstyle = {
      width: '80%',
    };
    return (
      <div className="col s12 m7 l12">
        <h6 className="header">
          <span className={paragraphstyle} >
            {this.props.data.author} : {this.props.data.title}
          </span>
        </h6>
        <div className="card horizontal">
          <div className="card-image">
            <img className={imgStyle} src={this.props.data.urlToImage} alt="No news image" />
          </div>
          <div className="card-stacked">
            <div className="card-content">
              <p className={paragraphstyle}>{this.props.data.description}</p>
            </div>
            <div className="card-action">
              <span>Published on {this.props.data.publishedAt} </span>
              <a href={this.props.data.url} target="_blank">Read on site </a>
            </div>
          </div>
        </div>
      </div>

    );
  }
}
const test = new Headlines();
// window.scrape = test.scrape();
export default Headlines;
