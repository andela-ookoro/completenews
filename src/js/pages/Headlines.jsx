import React from 'react';
// import Request from 'request';
// import Fs from 'fs';
import SortBY from './headlines/SortBy';
import Article from './Article';
import SourceOptions from './headlines/SourceOptions';
import NotifyStore from '../store/NotifyStore';
import SourceAction from '../action/sourceAction';
import Sources from '../store/SourceStore';
import * as HeadlineAction from '../action/headlineAction';
import HeadlineStore from '../store/HeadlineStore';
import AuthStore from '../store/authStore';


class Headlines extends React.Component {
  constructor() {
    super();
    this.state = {
      source: '',
      sort: '',
      sources: [],
      articles: [],
      articleSource: '',
      sortBy: [],
      currentSort: '',
      categories: [],
      message: '',
      isAuth: (!JSON.parse(localStorage.getItem('cat')) === {}),
      isDb: false,
      scrapeUrl: '',
    };
    this.count = 0;
    this.fecthHealines = this.fecthHealines.bind(this);
    this.getSources = this.getSources.bind(this);
    this.toTitleCase = (str =>
       str.replace(/-/g, ' ').replace(/\w\S*/g, txt =>
        txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase())
      );
    this.fetchAvailableSort = this.fetchAvailableSort.bind(this);
    this.notifyUser = this.notifyUser.bind(this);
    this.sourceChange = this.sourceChange.bind(this);
    this.headlineChange = this.headlineChange.bind(this);
    this.dbheadlineChange = this.dbheadlineChange.bind(this);
    this.authChange = this.authChange.bind(this);
    this.scrape = ((e) => {
      e.preventDefault();
      const index = e.target.value;
      const articles = JSON.parse(localStorage.articles);
      const article = articles[index];
      const scrapeUrl = article.url.toString();
      const scrapeTitle = article.title;
      if (scrapeUrl.includes('https')) {
        this.setState({
          scrapeUrl: '',
          message: 'Cannot view page; access blocked by source',
        });
      } else {
        this.setState({ message: '', scrapeUrl, scrapeTitle });
      }
      // Request(url, (error, response, html) => {
      //   if (!error) {
      //     Fs.writeFile('output.html', html, (err) => {
      //       console.log(`File successfully written! - Check your project
      //             directory for the output.json file`);
      //     });
      //   }
      // });
    });
    this.viewFavourite = (() => {
      let userEmail = JSON.parse(localStorage.getItem('userProfile'))
                      .email.toString().replace('.', '_');
      userEmail = userEmail.substring(0, userEmail.indexOf('@'));
      HeadlineAction.getDbHeadlines(userEmail);
      this.setState({ scrapeUrl: '' });
    });
    this.resetScrapeUrl = this.resetScrapeUrl.bind(this);
  }
  // this method runs before the component render it content
  componentWillMount() {
    if (this.state.isAuth) {
      this.viewFavourite();
    }
    localStorage.getItem('sources');
    this.getSources();
    HeadlineStore.on('dbchange', this.dbheadlineChange);
    HeadlineStore.on('change', this.headlineChange);
    AuthStore.on('change', this.authChange);
    Sources.on('change', this.sourceChange);
    NotifyStore.on('change', this.notifyUser);
    const userinfo = JSON.parse(localStorage.getItem('userProfile'));
    if (userinfo) {
      this.setState({ isAuth: true });
    }
  }

  componentWillUnmount() {
    HeadlineStore.removeListener('change', this.headlineChange);
    HeadlineStore.removeListener('dbchange', this.dbheadlineChange);
    Sources.removeListener('change', this.getSources);
    NotifyStore.removeListener('change', this.notifyUser);
    AuthStore.removeListener('change', this.authChange);
  }

  getSources() {
    if (!localStorage.getItem('sources')) {
      SourceAction();
    } else {
      this.setState(
        {
          sources: JSON.parse(localStorage.getItem('sources')),
          categories: JSON.parse(localStorage.getItem('cat')),
        });
    }
  }

  dbheadlineChange() {
    const headlines = HeadlineStore.headlines;
    const error = HeadlineStore.error;
    localStorage.setItem('articles', JSON.stringify(headlines));
    this.setState({
      articles: headlines,
      message: error,
      articleSource: 'Favourite Headlines',
      sortBy: [],
      currentSort: '',
      isAuth: false,
      scrapeUrl: '',
    });
  }

  headlineChange() {
    const headlines = HeadlineStore.headlines;
    const error = HeadlineStore.error;
    const userinfo = JSON.parse(localStorage.getItem('userProfile'));
    localStorage.setItem('articles', JSON.stringify(headlines));
    this.setState({
      articles: headlines,
      message: error,
      isAuth: (userinfo !== {}),
      scrapeUrl: '',
    });
  }

  sourceChange() {
    const sources = Sources.sources;
    const sourcescategories = {};
    const categories = [];
    sources.forEach((source) => {
      if (!sourcescategories.hasOwnProperty(source.category)) {
        sourcescategories[source.category] = [];
        categories.push(source.category);
      }
      sourcescategories[source.category].push(source);
    });
    localStorage.setItem('cat', JSON.stringify(categories));
    localStorage.setItem('categories', JSON.stringify(sourcescategories));
    localStorage.setItem('sources', JSON.stringify(sources));
    this.setState({ sources, categories });
  }

  authChange() {
    this.setState({ isAuth: AuthStore.isAuth });
    // console.log('auth');
  }

  notifyUser() {
    this.setState({ message: NotifyStore.message });
  }

  // get available sort parameter
  fetchAvailableSort(e) {
    e.preventDefault();
    let cursource = e.target.getAttribute('value');
    if (!cursource) {
      cursource = e.target.value;
    }
    // fix source name
    const sourceName = this.toTitleCase(cursource.toString());
    const sources = JSON.parse(localStorage.sources);
    const sourceNode = sources.filter(obj => obj.id === cursource);
    HeadlineAction.getHeadlines(cursource, '');
    this.setState({
      sources,
      source: cursource,
      articleSource: sourceName,
      sortBy: (sourceNode[0].sortBysAvailable.length > 1) ?
      sourceNode[0].sortBysAvailable : [],
      scrapeUrl: '',
    });
  }
   // fetct headlines
  fecthHealines(e) {
    e.preventDefault();
    const sort = e.target.value;
    const source = this.state.source;
    HeadlineAction.getHeadlines(source, sort);
    this.setState({
      currentSort: sort,
      scrapeUrl: '',
    });
  }

  resetScrapeUrl() {
    this.setState({ scrapeUrl: '' });
  }


  render() {
    return (
      <div className="row">
        <div className="col s2" id="side-nav">
          Sources
          <select
            name="sources" size="25" id="sources"
            className={'browser-default'} onChange={this.fetchAvailableSort}
          >
            {(this.state.sources.length < 1) ?
              ''
              :
            this.state.sources.map(source =>
              <option
                key={source.id} value={source.id} title={source.description}
              >
                {source.name}
              </option>,
            )}
          </select>
          <ul className="collapsible" data-collapsible="accordion">
            {(this.state.categories.length < 1 || !this.state.categories) ?
              ''
              :
            this.state.categories.map(cat =>
              <li key={cat}>
                <div className="collapsible-header"> {this.toTitleCase(cat)}
                </div>
                <div className="collapsible-body">
                  {JSON.parse(localStorage.getItem('categories'))[cat].map(
                    source =>
                      <SourceOptions
                        key={source.id} name={source.name} title={source.description}
                        id={source.id} fetchAvailableSort={this.fetchAvailableSort}
                      />,
                  )}
                </div>
              </li>,
            )}
          </ul>
        </div>
        <div className={'col s10'} id="articles">
          <div id="articles-menu">
            <div >
              {(this.state.message === '' && this.state.articles.length < 1)
                ?
                  <div className="progress">
                    <div className="indeterminate" />
                  </div>
                :
                  <h5> {this.state.message} </h5>
              }
            </div>
            <h5 name={'source'}>
              {this.state.articleSource}
              {
                (this.state.currentSort === '') ?
                ' '
                :
                ` ${this.toTitleCase(this.state.currentSort)}   
                ${this.state.articles.length}  Headlines `
              }
              {this.state.sortBy.map((sortBy, i) =>
                <SortBY
                  key={i} data={sortBy} source={this.state.source}
                  onClick={this.fecthHealines}
                />,
                )}
            </h5>
          </div>
          <div id="articles">
            {(this.state.scrapeUrl) ?
              <div className="scapediv">
                <h6 className="header"> {this.state.scrapeTitle}
                  <button onClick={this.resetScrapeUrl}>
                     &larr; View more Headlines </button>
                </h6>
                <iframe src={this.state.scrapeUrl} />
              </div>
            :
              this.state.articles.map((article, i) =>
                <Article
                  key={i} id={i} author={article.author} title={article.title}
                  urlToImage={article.urlToImage} description={article.description}
                  publishedAt={article.publishedAt} url={article.url}
                  source={(article.source) ? article.source : this.state.source}
                  isAuth={this.state.isAuth} scrape={this.scrape}
                />,
              )
            }
          </div>
        </div>
      </div>
    );
  }
}

export default Headlines;
