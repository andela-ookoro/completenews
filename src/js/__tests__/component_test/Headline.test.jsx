import React from 'react';
import { mount } from 'enzyme';
import Renderer from 'react-test-renderer';
import Headlines from '../../pages/Headlines';
import Visit from '../../helpers/visit';

const sources = [
  {
    "id":"abc-news-au",
    "name":"ABC News (AU)",
    "description":"Australia's most trusted source of local, national and world news. Comprehensive, independent, in-depth analysis, the latest business, sport, weather and more.",
    "url":"http://www.abc.net.au/news",
    "category":"general",
    "language":"en",
    "country":"au",
    "urlsToLogos":{"small":"","medium":"","large":""},
    "sortBysAvailable":["top"]
  },
  {
    "id":"al-jazeera-english",
    "name":"Al Jazeera English",
    "description":"News, analysis from the Middle East and worldwide, multimedia and interactives, opinions, documentaries, podcasts, long reads and broadcast schedule.",
    "url":"http://www.aljazeera.com",
    "category":"general",
    "language":"en",
    "country":"us",
    "urlsToLogos":{"small":"","medium":"","large":""},
    "sortBysAvailable":["top","latest"]
  },
];
// localStorage.setItem('sources', JSON.stringify(sources));
// localStorage.setItem('categories', JSON.stringify(['general', 'sport']));

let originalTimeout;
beforeEach(() => {
  originalTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL;
  jasmine.DEFAULT_TIMEOUT_INTERVAL = 20000;
});

afterEach(() => {
  jasmine.DEFAULT_TIMEOUT_INTERVAL = originalTimeout;
});

describe('End to End test ', () => {
  it('Article menu should display name of source when user select a source', 
  (done) => {
    const NEWTODO_INPUT = '#sources';
    Visit('/')
    .refresh()
      .wait(NEWTODO_INPUT)
      // type a todo and press ENTER
      .select(NEWTODO_INPUT, 'abc-news-au')
      .wait('#articles-menu')
      .evaluate(() => {
        return document.querySelector('#articles-menu').innerText;
      })
      .then((res) => {
        console.log(res);
        expect(res).toMatch('Abc News Au');
        done();
      })
      .catch((error) => {
        console.log(error);
      });
  });
  
    it('Article menu should display name of source', (done) => {
    const NEWTODO_INPUT = '#sources';
    Visit('/')
    .refresh()
      .wait(NEWTODO_INPUT)
      // type a todo and press ENTER
      .select(NEWTODO_INPUT, 'abc-news-au')
      .wait('#articles-menu')
      .evaluate(() => {
        return document.querySelector('#articles-menu').innerText;
      })
      .then((res) => {
        console.log(res);
        expect(res).toMatch('Abc News Au');
        done();
      })
      .catch((error) => {
        console.log(error);
      });
  });
});

test('Component render the article template', () => {
  const component = Renderer.create(
    <Headlines />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
