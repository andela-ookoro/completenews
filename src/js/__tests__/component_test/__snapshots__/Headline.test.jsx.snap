// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component render the article template 1`] = `
<div
  className="row"
>
  <div
    className="col s2"
    id="side-nav"
  >
    Sources
    <select
      className="browser-default"
      id="sources"
      name="sources"
      onChange={[Function]}
      size="25"
      style={
        Object {
          "height": "10em",
        }
      }
    >
      
    </select>
    <ul
      className="collapsible"
      data-collapsible="accordion"
    >
      
    </ul>
  </div>
  <div
    className="col s10"
    id="articles"
  >
    <div
      id="articles-menu"
    >
      <div>
        <div
          className="progress"
        >
          <div
            className="indeterminate"
          />
        </div>
      </div>
      <h5
        name="source"
      >
        
         
      </h5>
    </div>
    <div
      id="articles"
    />
  </div>
</div>
`;
